Sources/Build/BuildPlan/BuildPlan.swift

```diff
@@ -692,6 +727,203 @@ public class BuildPlan: SPMBuildCore.BuildPlan {
+                    observabilityScope: observability.topScope
+                )
+            } else {
+                pluginDerivedSources = .init(paths: [], root: package.path)
+                pluginDerivedResources = []
+            }
+
+            let result = try temp_await {
+                pluginModule.invoke(
+                    module: plugin,
+                    action: .createBuildToolCommands(
+                        package: package,
+                        target: module,
+                        pluginGeneratedSources: pluginDerivedSources.paths,
+                        pluginGeneratedResources: pluginDerivedResources.map(\.path)
+                    ),
+                    buildEnvironment: buildParameters.buildEnvironment,
+                    scriptRunner: configuration.scriptRunner,
+                    workingDirectory: package.path,
+                    outputDirectory: pluginOutputDir,
+                    toolSearchDirectories: [buildParameters.toolchain.swiftCompilerPath.parentDirectory],
+                    accessibleTools: accessibleTools,
+                    writableDirectories: writableDirectories,
+                    readOnlyDirectories: readOnlyDirectories,
+                    allowNetworkConnections: [],
+                    pkgConfigDirectories: pkgConfigDirectories,
+                    sdkRootPath: buildParameters.toolchain.sdkRootPath,
+                    fileSystem: fileSystem,
+                    modulesGraph: modulesGraph,
+                    observabilityScope: observabilityScope,
+                    completion: $0
+                )
+            }
+
+            if surfaceDiagnostics {
+                let diagnosticsEmitter = observabilityScope.makeDiagnosticsEmitter {
+                    var metadata = ObservabilityMetadata()
+                    metadata.moduleName = module.name
+                    metadata.pluginName = result.plugin.name
+                    return metadata
+                }
+
+                for line in result.textOutput.split(whereSeparator: { $0.isNewline }) {
+                    diagnosticsEmitter.emit(info: line)
+                }
+
+                for diag in result.diagnostics {
+                    diagnosticsEmitter.emit(diag)
+                }
+            }
+
+            // Add a BuildToolPluginInvocationResult to the mapping.
+            buildToolPluginResults.append(result)
+        }
+
+        return buildToolPluginResults
+    }
+
+    /// Runs any prebuild commands associated with the given list of plugin invocation results,
+    /// in order, and returns the results of running those prebuild commands.
+    fileprivate static func runPrebuildCommands(
+        using pluginConfiguration: PluginConfiguration,
+        for pluginResults: [BuildToolPluginInvocationResult],
+        fileSystem: any FileSystem,
+        observabilityScope: ObservabilityScope
+    ) throws -> [PrebuildCommandResult] {
```

Comment:
Probably worth renaming the type too, to make it more obvious which feature it's referring to.

```suggestion
    ) throws -> [CommandPluginResult] {
```

Reply:
Renamed!